eng/pipelines/pr.yml

```diff
@@ -0,0 +1,241 @@
+parameters:
+- name: RunUnitTestsOnWindows
+  displayName: Run unit tests on Windows
+  type: boolean
+  default: true
+- name: RunFunctionalTestsOnWindows
+  displayName: Run functional tests on Windows
+  type: boolean
+  default: true
+- name: RunUnitTestsOnLinux
+  displayName: Run unit tests on Linux
+  type: boolean
+  default: true
+- name: RunFunctionalTestsOnLinux
+  displayName: Run functional tests on Linux
+  type: boolean
+  default: true
+- name: RunUnitTestsOnMacOS
+  displayName: Run unit tests on MacOS
+  type: boolean
+  default: true
+- name: RunFunctionalTestsOnMacOS
+  displayName: Run functional tests on MacOS
+  type: boolean
+  default: true
+
+variables:
+  DOTNET_NOLOGO: 1
+  NUGET_EXPERIMENTAL_CHAIN_BUILD_RETRY_POLICY: 3,1000
+  SkipSigning: true
+  CI: true
+
+trigger:
+  branches:
+    include:
+    - dev
+  paths:
+    exclude:
+    - '*.md'
+
+pr:
+  branches:
+    include:
+    - dev
+  paths:
+    exclude:
+    - '*.md'
+
+stages:
+- ${{ if eq(parameters.RunUnitTestsOnWindows, true) }}:
+  - stage:
+    displayName: Unit Tests on Windows (.NET Framework 4.7.2)
+    dependsOn: []
+    jobs:
+    - template: pr.job.yml
+      parameters:
+        displayName: Unit Tests on Windows (.NET Framework 4.7.2)
+        osName: Windows
+        agentPool: NetCore-Public
+        agentDemands: ImageOverride -equals Windows.VS2022Preview.Amd64.Open
+        testType: Unit
+        testTargetFramework: net472
+
+- ${{ if eq(parameters.RunUnitTestsOnWindows, true) }}:
+  - stage:
+    displayName: Unit Tests on Windows (.NET 8.0)
+    dependsOn: []
+    jobs:
+    - template: pr.job.yml
+      parameters:
+        displayName: Unit Tests on Windows (.NET 8.0)
+        osName: Windows
+        agentPool: NetCore-Public
+        agentDemands: ImageOverride -equals Windows.VS2022Preview.Amd64.Open
+        testType: Unit
+        testTargetFramework: net8.0
+
+- ${{ if eq(parameters.RunUnitTestsOnWindows, true) }}:
+  - stage:
+    displayName: Unit Tests on Windows (.NET Core 3.1)
+    dependsOn: []
+    jobs:
+    - template: pr.job.yml
+      parameters:
+        displayName: Unit Tests on Windows (.NET Core 3.1)
+        osName: Windows
+        agentPool: NetCore-Public
+        agentDemands: ImageOverride -equals Windows.VS2022Preview.Amd64.Open
+        testType: Unit
+        testTargetFramework: netcoreapp3.1
+
+- ${{ if eq(parameters.RunFunctionalTestsOnWindows, true) }}:
+  - stage:
+    displayName: Functional Tests on Windows (.NET Framework 4.7.2)
+    dependsOn: []
+    jobs:
+    - template: pr.job.yml
+      parameters:
+        displayName: Functional Tests on Windows (.NET Framework 4.7.2)
+        osName: Windows
+        agentPool: NetCore-Public
+        agentDemands: ImageOverride -equals Windows.VS2022Preview.Amd64.Open
+        testType: Functional
+        testTargetFramework: net472
```

Comment:
I'm not a fan of testing each TFM in a separate job. I know different people have different tolerances for waiting, but I will find this pipeline breaking when we change target frameworks more annoying than having CI jobs take a little longer  (but there will be fewer of them too). I always go do something else while I wait for CI to finish.

Reply:
The main advantages of having them fanned out is more parallelism (the machines we get for Linux and Mac have 2 cores, Windows machines have 4) and you can re-run a smaller stage in case of a flaky test.  I'm more than happy to tweak this configuration down the road but I believe this is the best balance for the current state of things.